File: GOF/抽象工厂AbstractFactory/java/src/factory/MazeFactory.java
package factory;
import maze.Door;
import maze.Maze;
import maze.Room;
import maze.Wall;


public abstract class MazeFactory {
    public Maze makeMaze()
    {
        return new Maze();
    }

    public Wall makeWall()
    {
        return new Wall();
    }

    public Room makeRoom(int n)
    {
        return new Room(n);
    }

    public Door makeDoor(Room r1, Room r2)
    {
        return new Door(r1, r2);
    }
}

File: GOF/抽象工厂AbstractFactory/cpp/factory/mazefactory.cpp
//
// Created by gaojian on 19-2-27.
//

#include "mazefactory.h"
#include "../maze/maze.h"
#include "../maze/wall.h"
#include "../maze/door.h"
#include "../maze/room.h"

using std::make_shared;

std::shared_ptr<Maze> MazeFactory::makeMaze() const
{
	return make_shared<Maze>();
}

std::shared_ptr<Wall> MazeFactory::makeWall() const
{
	return make_shared<Wall>();
}

std::shared_ptr<Room> MazeFactory::makeRoom(int n) const
{
	return make_shared<Room>(n);
}

std::shared_ptr<Door> MazeFactory::makeDoor(std::weak_ptr<Room> r1, std::weak_ptr<Room> r2) const
{
	return make_shared<Door>(r1, r2);
}